src/java.base/macosx/native/libnio/fs/MacOSXWatchService.c

```diff
@@ -0,0 +1,237 @@
+                                               javaEventPathsArray);
+        }
+
+        callJavaCallback(env,
+                         ptr_to_jlong(streamRef),
+                         javaEventPathsArray,
+                         ptr_to_jlong(&eventFlags[eventIndex]));
+
+        if ((*env)->ExceptionCheck(env)) {
+            (*env)->ExceptionDescribe(env);
+        }
+
+        if (localFramePushed) {
+            (*env)->PopLocalFrame(env, NULL);
+        }
+
+        eventIndex += numEventsToReport;
+    }
+}
+
+/**
+ * Creates a new FSEventStream and returns FSEventStreamRef for it.
+ */
+JNIEXPORT jlong JNICALL
+Java_sun_nio_fs_MacOSXWatchService_eventStreamCreate(JNIEnv* env,
+                                                     __unused jclass clazz,
+                                                     jstring dir,
+                                                     jdouble latencyInSeconds,
+                                                     jint flags)
+{
+    const CFStringRef path = toCFString(env, dir);
+    CHECK_NULL_RETURN(path, 0);
+    const CFArrayRef pathsToWatch = CFArrayCreate(NULL, (const void **) &path,
+                                                  1, NULL);
+    CHECK_NULL_RETURN(pathsToWatch, 0);
+
+    const FSEventStreamRef stream = FSEventStreamCreate(
+            NULL,
+            &callback,
+            NULL,
+            pathsToWatch,
+            kFSEventStreamEventIdSinceNow,
+            (CFAbsoluteTime) latencyInSeconds,
+            flags);
+
+    return ptr_to_jlong(stream);
+}
+
+
+/**
+ * Schedules the given FSEventStream with the given run loop. Starts the stream
+ * so that the run loop can receive events from the stream.
+ */
+JNIEXPORT void JNICALL
+Java_sun_nio_fs_MacOSXWatchService_eventStreamSchedule(__unused JNIEnv* env,
+                                                       __unused jclass clazz,
+                                                       jlong eventStreamRef,
+                                                       jlong runLoopRef)
+{
+    const FSEventStreamRef stream  = jlong_to_ptr(eventStreamRef);
+    const CFRunLoopRef     runLoop = jlong_to_ptr(runLoopRef);
+
+    FSEventStreamScheduleWithRunLoop(stream, runLoop, kCFRunLoopDefaultMode);
+    FSEventStreamStart(stream);
+}
+
+/**
+ * Performs the steps necessary to dispose of the given FSEventStreamRef.
+ * The stream must have been started and scheduled with a run loop.
+ */
+JNIEXPORT void JNICALL
+Java_sun_nio_fs_MacOSXWatchService_eventStreamStop(__unused JNIEnv* env,
+                                                   __unused jclass clazz,
+                                                   jlong eventStreamRef)
+{
+    const FSEventStreamRef streamRef = jlong_to_ptr(eventStreamRef);
+
+    FSEventStreamStop(streamRef);       // Unregister with the FS Events service.
+                                        // No more callbacks from this stream.
```

Comment:
Does `FSEventStreamUnscheduleFromRunLoop()` need to be called here?

Reply:
I think not. The [documentation](https://developer.apple.com/documentation/coreservices/1446990-fseventstreaminvalidate?language=objc) for `FSEventStreamInvalidate()` that gets called immediately after this line states the following:
> Invalidates the stream, like CFRunLoopSourceInvalidate() does for a CFRunLoopSourceRef. It will be unscheduled  from any runloops or dispatch queues upon which it had been scheduled.

So calling `FSEventStreamUnscheduleFromRunLoop()` seems superfluous; its semantics is described as
> ... removes the stream from the specified run loop, like CFRunLoopRemoveSource() does for a CFRunLoopSourceRef.

Reply:
That sounds correct.